### ISSUE TITLE
pamd.py - Module more than 2 parameter on a rule.control,  duplicate parameters each time you run the playbook

### ISSUE THREAD

**Issue by silvinux** [ansible/ansible#25522]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME

/usr/lib/python2.7/site-packages/ansible/modules/system/pamd.py

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /var/lib/ansible/local/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.5 (default, Aug 29 2016, 10:12:21) [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
```

##### CONFIGURATION
[defaults]
inventory      = ./inventory_dlabtuxm
forks          = 30
remote_port    = 22
gather_subset	= all
gather_timeout	= 10
roles_path	= ./roles
host_key_checking = False
timeout = 5
remote_user = root 
log_path = /var/log/ansible/ansible.log
ansible_managed = Ansible managed
squash_actions = apk,apt,dnf,homebrew,package,pacman,pkgng,yum,zypper
no_log = False
var_compression_level = 9
module_compression = 'ZIP_DEFLATED'
max_diff_size = 1048576
[privilege_escalation]
become=True
become_method=sudo
become_user=root
become_ask_pass=False
[ssh_connection]
[colors]

##### OS / ENVIRONMENT

N/A
##### SUMMARY

When you have more than 2 parameters on a rule.control, the module duplicate parameters starting at second parameter, as many times as rule_types exits before. No matter whether apply some changes or not (in the playbook), this happens all the time and inserts these duplicates parameters each time you run the playbook.
##### STEPS TO REPRODUCE

Example: /etc/pam.d/system-auth

auth      	required	pam_env.so 
auth      	required	pam_tally2.so deny=5 onerr=fail
auth      	sufficient	pam_fprintd.so 
auth      	[default=1 success=ok]	pam_localuser.so 
auth      	[success=done ignore=ignore default=die] ignore=ignore default=die] ignore=ignore default=die] ignore=ignore default=die] ignore=ignore default=die] pam_unix.so nullok

As you can see, there are 4 rule_types (AUTH), before the rule_control ([success=done ignore=ignore default=die]), and it duplicates, starting the second parameter, a total of 4 times.

I starting to analyze the code, and I think the string split to assign values to each rule could be enhanced with a loop. That way no matter how many parameters rule.control have.

  @classmethod
    def rulefromstring(cls, stringline):
        split_line = stringline.split()

        rule_type = split_line[0]
        rule_control = split_line[1]

        if rule_control.startswith('['):
            rule_control = stringline[stringline.index('['):
                                      stringline.index(']')+1]
###################################################################
        if "]" in split_line[2]:
            rule_module_path = split_line[3]
            rule_module_args = split_line[4:]
        else:
            rule_module_path = split_line[2]
            rule_module_args = split_line[3:]
###################################################################

        return cls(rule_type, rule_control, rule_module_path, rule_module_args)

I tried that piece of code, for example with this line:

s="session      [success=1 default=ignore]      pam_succeed_if.so service in crond quiet use_uid"

The result is OK:

python test_rulefromstring.py 
('session', '[success=1 default=ignore]', 'pam_succeed_if.so', ['service', 'in', 'crond', 'quiet', 'use_uid'])
 
But if you try with more than two parameter in the rule_control:

s="account      [success=ok new_authtok_reqd=ok ignore=ignore default=bad] pam_sss.so service in crond quiet use_uid"

The result is:

python test_rulefromstring.py 
('account', '[success=ok new_authtok_reqd=ok ignore=ignore default=bad]', 'new_authtok_reqd=ok', ['ignore=ignore', 'default=bad]', 'pam_sss.so', 'service', 'in', 'crond', 'quiet', 'use_uid'])

This means that rule_module_path and rule_module_args get an incorrect value.

The loop:

for line, val in enumerate(stringline.split()):
      print line, val
      if "]" in val:
        rule_module_path = split_line[line+1]
        rule_module_args = split_line[line+2:]
      else:
        rule_module_path = split_line[2]
        rule_module_args = split_line[3:]

# python test_rulefromstring.py  
0 account
1 [success=ok
2 new_authtok_reqd=ok
3 ignore=ignore
4 default=bad]
5 pam_sss.so
6 service
7 in
8 crond
9 quiet
10 use_uid
('account', '[success=ok new_authtok_reqd=ok ignore=ignore default=bad]', 'pam_sss.so', ['service', 'in', 'crond', 'quiet', 'use_uid'])
 
I haven't been able to found which piece of code change the rule, would you be so kind to help me or guide me on this.

Thanks in advance,

```yaml

```

##### EXPECTED RESULTS

Example: /etc/pam.d/system-auth

auth      	required	pam_env.so 
auth      	required	pam_tally2.so deny=5 onerr=fail
auth      	sufficient	pam_fprintd.so 
auth      	[default=1 success=ok]	pam_localuser.so 
auth      	[success=done ignore=ignore default=die]  pam_unix.so nullok

##### ACTUAL RESULTS

Example: /etc/pam.d/system-auth

auth      	required	pam_env.so 
auth      	required	pam_tally2.so deny=5 onerr=fail
auth      	sufficient	pam_fprintd.so 
auth      	[default=1 success=ok]	pam_localuser.so 
auth      	[success=done ignore=ignore default=die] ignore=ignore default=die] ignore=ignore default=die] ignore=ignore default=die] ignore=ignore default=die] pam_unix.so nullok

```

```

**Comment by jctanner** [comment]
cc @kevensen

**Comment by kevensen** [comment]
@jctanner Go ahead and assign this to me.  I'll take a look at it this weekend.

**Comment by kevensen** [comment]
@jctanner @silvinux I haven't forgotten about this.  I just haven't had a chance to look at it yet.

**Comment by kevensen** [comment]
@silvinux Can you please provide the Ansible task that you used when you discovered the bug?

**Comment by silvinux** [comment]
@jctanner It will duplicate the last two rule_control parameters (ignore=ignore default=die]) as many tasks run in the playbook.
```
---
- hosts: lab
  gather_facts: yes
  remote_user: gnfux
  become: yes
  become_method: sudo
  tasks:
    - name: "Ensure content of /etc/pam.d/system-auth"
      pamd:
        path: /etc/pam.d/
        name: system-auth
        type: password
        control: requisite
        module_path: pam_cracklib.so
        new_type: password
        new_control: sufficient
        new_module_path: pam_unix.so
        module_arguments: 'remember=6 sha512 shadow nullok try_first_pass use_authtok'
        state: after

# ansible-playbook -i hosts.yml test.yml 

PLAY [lab] *********************************************************************

TASK [setup] *******************************************************************
ok: [labtux]

TASK [Ensure content of /etc/pam.d/system-auth] ********************************
ok: [labtux]

PLAY RECAP *********************************************************************
dlabtuxm4                  : ok=2    changed=0    unreachable=0    failed=0   


# head /etc/pam.d/system-auth
#%PAM-1.0
# This file is auto-generated.
# User changes will be destroyed the next time authconfig is run.
auth      	required	pam_env.so 
auth      	required	pam_tally2.so deny=5 onerr=fail
auth      	sufficient	pam_unix.so nullok try_first_pass
auth      	[success=done ignore=ignore default=die]	ignore=ignore default=die] pam_unix.so nullok
auth      	requisite	pam_succeed_if.so uid >= 1000 quiet_success
auth      	required	pam_deny.so `
```

**Comment by dsmorse** [comment]
I'm seeing the same error with the pamd module.  My play looks like this:

> - name: "MEDIUM | RHEL-07-010200 | PATCH | The PAM system service must be configured to store only encrypted representations of passwords."
  pamd:
      name: system-auth
      state: "{{ item.state }}"
      type: password
      control: sufficient
      module_path: pam_unix.so
      module_arguments: "{{ item.args }}"
  with_items:
      - state: args_present
        args:
          - "sha512"
      - state: args_absent
        args:
          - "md5"
          - "bigcrypt"
          - "sha256"
          - "blowfish"


If the box has been realm joined first then the initial /etc/pam.d/system-auth looks like:

> auth        required      pam_env.so
   auth        [default=1 success=ok] pam_localuser.so
   auth        [success=done ignore=ignore default=die] pam_unix.so try_first_pass
   auth        requisite     pam_succeed_if.so uid >= 1000 quiet_success
   auth        sufficient    pam_sss.so forward_pass
   auth        required      pam_deny.so


After executing the pamd module (and not updating the pam_localuser.so or the pam_unix.so line) my file looks like this:

> auth        required        pam_env.so
   auth            [default=1 success=ok]  pam_localuser.so
   auth            [success=done ignore=ignore default=die]        ignore=ignore default=die] ignore=ignore default=die] pam_unix.so try_first_pass
   auth            requisite       pam_succeed_if.so uid >= 1000 quiet_success
   auth            sufficient      pam_sss.so forward_pass
   auth            required        pam_deny.so
...

**Comment by kevensen** [comment]
@dsmorse @silvinux Thank you for the input.  I'm taking a look at this now.

**Comment by kevensen** [comment]
@silvinux @dsmorse If you have time, do you mind testing this pull request in your environment?  I changed the mechanism by which the rule was being parsed.

**Comment by dsmorse** [comment]
I can confirm in my testing it works. Thanks

**Comment by silvinux** [comment]
Works like a charm!!!

**Comment by silvinux** [comment]
Now duplicates rule_module_args (onerr=fail in this case)

auth      	required	pam_tally2.so deny=5 onerr=fail onerr=fail

**Comment by kevensen** [comment]
@silvinux could you please share your play?

**Comment by silvinux** [comment]
My bad!!! :(, I had two tasks state: after and  state: args_present, seems to be that way duplicate the parameter, I just keep one task. Thanks

---
- hosts: lab
  gather_facts: yes
  remote_user: gnfux
  become: yes
  become_method: sudo
  tasks:
    - name: "Ensure /etc/pam.d/system-auth has changed #auth required pam_env.so# to #auth required pam_tally2 deny=5 onerr=fail#"
      become: yes
      pamdl:
        path: /etc/pam.d/
        name: system-auth
        type: auth
        control: required
        module_path: pam_env.so
        new_type: auth
        new_control: required
        new_module_path: pam_tally2.so
        module_arguments: 'deny=5 onerr=fail'
        state: after

    - name: "Ensure correct parameters for pam_tally2 --> deny=5 onerr=fail#"
      become: yes
      pamdl:
        path: /etc/pam.d/
        name: system-auth
        type: auth
        control: required
        module_path: pam_tally2.so
        module_arguments: 'deny=5 onerr=fail'
        state: args_present

**Comment by kevensen** [comment]
No worries!

**Comment by silvinux** [comment]
Hello, It's me again!!!. It seems when you have equal, less or more than symbols, will ignore the parameter, and I even tried to change that. 

Original file.
```
[root@labtux local]# grep uid /etc/pam.d/system-auth
auth        requisite     pam_succeed_if.so uid >= 1025 quiet_success
account     sufficient    pam_succeed_if.so uid < 1025 quiet
session     [success=1 default=ignore] pam_succeed_if.so service in crond quiet use_uid
```

After pamd applied.
```
[root@labtux local]# grep uid /etc/pam.d/system-auth
auth            requisite       pam_succeed_if.so uid
account         sufficient      pam_succeed_if.so uid
session         [success=1 default=ignore]      pam_succeed_if.so service in crond quiet use_uid
```


```
---
- hosts: lab
  gather_facts: yes
  remote_user: gnfux
  become: yes
  become_method: sudo
  tasks:
    - name: "Ensure /etc/pam.d/system-auth has changed #auth required pam_env.so# to #auth required pam_tally2 deny=5 onerr=fail#"
      become: yes
      pamd:
        path: /etc/pam.d/
        name: system-auth
        type: auth
        control: required
        module_path: pam_env.so
        new_type: auth
        new_control: required
        new_module_path: pam_tally2.so
        module_arguments: 'deny=5 onerr=fail'
        state: updated
```

Thanks in advance,

**Comment by kevensen** [comment]
@silvinux I hadn't considered that case.  Just pushed a fix.  The only caveat is that there cant be any spaces around the comparison.

uid>=1025 
as opposed to 
uid >= 1025

**Comment by kevensen** [comment]
+label affects_2.3

**Comment by silvinux** [comment]
@kevensen sorry for the delay in action, the module doesn't work. I tried as #23963 suggest. Could you help on this?. 


`_failed: [labtux] (item=try_first_pass) => {"failed": true, "item": "try_first_pass", "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_tu8G1Z/ansible_module_pamd.py\", line 634, in <module>\r\n    main()\r\n  File \"/tmp/ansible_tu8G1Z/ansible_module_pamd.py\", line 569, in main\r\n    pamd = PamdService(path, service, module)\r\nTypeError: __init__() takes at most 2 arguments (4 given)\r\n", "msg": "MODULE FAILURE"}_`

testing.yml
```
---
- hosts: lab
  gather_facts: yes
  remote_user: gnfux
  become: yes
  become_method: sudo
  tasks:
    - name: hardening auth pam_unix.so
      pamdl:
        name: system-auth
        type: auth
        module_path: pam_unix.so
        control: sufficient
        module_arguments: "{{ item }}"
        state: args_present
      with_items:
        - nullok
        - try_first_pass

    - name: "Ensure /etc/pam.d/system-auth has changed #auth required pam_env.so# to #auth required pam_tally2 deny=5 onerr=fail#"
      become: yes
      pamdl:
        path: /etc/pam.d/
        name: system-auth
        type: auth
        control: required
        module_path: pam_env.so
        new_type: auth
        new_control: required
        new_module_path: pam_tally2.so
        module_arguments: "{{ item }}"
        state: updated
      with_items:
        - deny=5
        - onerr=fail
```


Thanks in advance,

**Comment by kevensen** [comment]
@silvinux Thanks for taking another look.  Have you tried the latest from https://github.com/ansible/ansible/pull/25817?

**Comment by silvinux** [comment]
@kevensen I downloaded the file from -->  Fix for #25522 ca2e393  7 days ago

**Comment by kevensen** [comment]
@silvinux Ah, yes.  There was a bug in that particular commit.  Can you please checkout  the current version?

**Comment by kevensen** [comment]
resolved_by_pr #25817

**Comment by silvinux** [comment]
@kevensen thanks, works OK, but I found another bug. Trying to change a parameter in a type, will be modified another one. and just keep two arguments modules. Thanks in advance,


play
```
---
- hosts: lab
  gather_facts: yes
  remote_user: gnfux
  become: yes
  become_method: sudo
  tasks:
	# Ensure content of /etc/pam.d/{system-auth,password-auth}
	- name: "pam_tally -- Ensure pam.d/{system-auth,password-auth} has changed #auth"
	  become: yes
	  pamdl:
	    path: /etc/pam.d/
	    name: "{{ item }}"
	    type: auth
	    control: required
	    module_path: pam_env.so
	    new_type: auth
	    new_control: required
	    new_module_path: pam_tally2.so
	    module_arguments: 'deny=5 onerr=fail'
	    state: after
	  with_items:
	    - system-auth
	    - password-auth
```

--- Origial file ---
```
[root@labtux ~]# cat /etc/pam.d/system-auth
#%PAM-1.0
# This file is auto-generated.
# User changes will be destroyed the next time authconfig is run.
auth        required      pam_env.so
auth        sufficient    pam_fprintd.so
auth        [default=1 success=ok] pam_localuser.so
auth        [success=done ignore=ignore default=die] pam_unix.so nullok try_first_pass
auth        requisite     pam_succeed_if.so uid >= 1025 quiet_success
auth        sufficient    pam_sss.so forward_pass
auth        required      pam_deny.so

password    requisite     pam_pwquality.so retry=5 minlen=6 dcredit=-1 ucredit=0 lcredit=-1 ocredit=0 reject_username  <=======
```



--- Modified file ---
```
[root@labtux ~]# cat /etc/pam.d/system-auth
#%PAM-1.0
# This file is auto-generated.
# User changes will be destroyed the next time authconfig is run.
auth      	required	pam_env.so
auth      	required	pam_tally2.so deny=5 onerr=fail
auth      	sufficient	pam_fprintd.so
auth      	[default=1 success=ok]	pam_localuser.so
auth      	[success=done ignore=ignore default=die]	pam_unix.so nullok try_first_pass
auth      	requisite	pam_succeed_if.so uid >= 1025 quiet_success
auth      	sufficient	pam_sss.so forward_pass
auth      	required	pam_deny.so

password  	requisite	pam_pwquality.so retry=5 minlen=6 dcredit=   <=======
```

**Comment by bcoca** [comment]
closed via #25817

**Comment by johnbianchi** [comment]
How can this be a "fix":
`` 
Just pushed a fix.  The only caveat is that there cant be any spaces around the comparison. 
 uid>=1025 as opposed to uid >= 1025
``

realm joining with SSSD writes the pam.d file, we do not have control if there are "spaces" around the comparison

before pamd module:
>  auth        requisite     pam_succeed_if.so uid >= 1000 quiet_success

after pamd module:
> auth      	requisite	pam_succeed_if.so uid

that is still broken
